lib/proxy/log_buffer/buffer.rb

```diff
@@ -0,0 +1,110 @@
+require 'date'
+require 'proxy/log_buffer/ring_buffer'
+
+# Adopted from Celluloid library (ring_buffer.rb).
+# Copyright (c) 2011-2014 Tony Arcieri. Distributed under the MIT License.
+# https://github.com/celluloid/celluloid/blob/0-16-stable/lib/celluloid/logging/ring_buffer.rb
+module Proxy::LogBuffer
+
+  LogRecord = Struct.new(:timestamp, :level, :message, :backtrace) do
+    def to_h
+      h = {}
+      self.class.members.each{|m| h[m.to_sym] = self[m]}
+      h[:level] = case h[:level]
+                  when ::Logger::Severity::INFO
+                    :INFO
+                  when ::Logger::Severity::WARN
+                    :WARN
+                  when ::Logger::Severity::ERROR
+                    :ERROR
+                  when ::Logger::Severity::FATAL
+                    :FATAL
+                  when ::Logger::Severity::DEBUG
+                    :DEBUG
+                  else
+                    :UNKNOWN
+                  end
+      h.delete(:backtrace) unless h[:backtrace]
+      h
+    end
+  end
+
+  class Buffer
+    def self.instance
+      @@buffer ||= Buffer.new
+    end
+
+    def initialize
+      @mutex = Mutex.new
+      @buffers_set_up = false
+      @failed_modules = {}
+    end
+
+    # late initialization
+    def setup_buffers(size = nil, size_tail = nil, level = nil, level_tail = nil)
```

Comment:
Because of settings. Logging is initialized before settings are loaded, apparently we want to log loading of settings. Therefore I am not able to initialize buffers in the constructor and I need late initialization after settings are loaded.

The workaround would be to create temporary buffer to store all messages at that early stage and then re-initialize and copy the records over to the final destination. I though it's not worth it. I can do it this way if you insist.